rfcs/0137-nix-language-version.md

```diff
@@ -0,0 +1,655 @@
+         - For example, one would have to batch language changes for a version bump to limit the number of increments
+           - (+) This would be beneficial for alternative implementations in terms of churn and effort to keep up
+       - Or implementors of alternative evaluators catching up with changes
+         - (+) Specifying the language precisely via the version actually offers alternative implementations an alternative to catching up: only support a given language version
+     - (+) This essentially nudges one to organise Nix language (specification or evaluator implementation ) development to be more independent of the rest of Nix
+       This is good, since it in turn forces stronger separation of concerns and more architectural clarity
+     - (-) Prohibits best-effort attempts at evaluating expressions with possibly incompatible evaluators
+       - (+) With the proposed level of strictness, one doesn't have to rely on best effort but can instead be explicit
+     - (-) Fixing evaluator bugs (i.e., clearly unintended behavior) after releases would technically require a version bump and therefore (theoretically) cooperation by expression authors
+       - This could be communicated with an increment in the Nix patch-level version, as is already practice
+
+   </details>
+
+   <details><summary>Alternatives</summary>
+
+   - Bump version only when evaluation result on prior version evaluators would be *substantially* different
+     - (+) Leaves room for judgement by developers
+       - (+) Allows controlling progression of versions to some degree
+     - (-) Conversely, leaves room for sneaking in breaking changes unannounced
+       - (-) This loses compatibility guarantees we'd get from a stricter paradigm
+       - (-) Deprives expression authors of ability to be selective with evaluator versions
+     - (-) Due to hashing this is often not much different from taking *any* change into account
+   - Bump version when prior evaluators would fail
+     - (-) Derivation hashes may change between evaluator versions if different evaluation results were allowed within one version
+       - This amounts to giving up on specifying the language exhaustively using a version label
+     - (+) Fewer version increments or precautions required given current development practice
+
+   </details>
+
+1. Whenever Nix drops support for evaluating prior language versions, a major version bump is required.
+
+   Example: Assuming the current language version is 8, the Nix release version is 2.20, and support for language version 6 is dropped, the next Nix release must be version 3.0
+
+   <details><summary>Arguments</summary>
+
+   - (+) This enforces that existing code that works will not break inadvertently when upgrading Nix
+
+   </details>
+
+1. Semantics are preserved across file boundaries for past language versions.
+
+   Example: [Best-effort interoperability](#best-effort-interoperability)
```

Comment:
While that section gives some examples, it's unclear to me whether we can come up with examples that would cause problems. I think the RFC should either give an extensive set of examples, or specify the full semantics of how such interoperability is achieved.

Reply:
To be honest, I'd rather crowd-source the pathological examples, but here's a hint at what the implementation could look like:

```suggestion
1. Semantics are preserved across file boundaries for past language versions.

   This should be fairly straightforward to implement since values passed around in the evaluator carry all the information needed to force them.
   Newer parts of the evaluator can always wrap their values in interfaces that are accepted by older parts, as far as possible.

   Example: [Best-effort interoperability](#best-effort-interoperability)
```

Reply:
> depending on where a function is defined it can unwittinly accept different types! 

Yes, since for given two language versions these are not the same function. I see the problem though: if you have tangled code paths like in the module system, introducing more moving parts will make reasoning about them a lot harder.

Reply:
> maybe a language version is too coarse and we should allow opting in to feature sets instead, with language versions merely being aliases for sets ...

Feature sets sound like a lot of cognitive and implementation overhead. Versions as aliases for feature sets sounds like only backward compatible changes. Which is of course sensible, but doesn't allow getting rid of baggage.

Reply:
I disagree with imposing even more cognitive overhead to regular users than already is done by this whole proposal, already on the grounds that it's hard enough for beginners as it is. I'd weigh end user and light expression author experience higher than comfort for those digging in the bowels of Nixpkgs. Not because these contributors are somehow less important, quite the opposite. But because a) doing bowel digging is a heavy investment regardless, and I suppose the worsening by additional complexity is at worst incremental b) we have to make sure that new contributors can get onboarded somewhat reliably, and we don't want to scare them away with even more moving parts.

The premise of this proposal is that the benefits of enabling controlled change to the language will outweigh the associated non-trivial costs. Agreement on adding versioning at all is, if nothing else, testament that changes to the language are desired.

Reply:
I also think that we should avoid specific feature flags. I think that in general moving the language forward is the goal. Sure, you can stay on an old version for a very long time but I think it makes sense to expect that once you do move forward you update to all of the latest changes. This avoids the need for the implementation to consider every combination of features (and possibly dependent features). It also avoids the user to need to keep a map of exactly what features are enabled when editing a file, having a linear scale of version is much easier to manager IMHO.